File: Pod/Client/SBTUITestTunnelClient.m
// SBTUITestTunnelClient.m
//
// Copyright (C) 2016 Subito.it S.r.l (www.subito.it)
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#if DEBUG
    #ifndef ENABLE_UITUNNEL 
        #define ENABLE_UITUNNEL 1
    #endif
#endif

#if ENABLE_UITUNNEL

#import "SBTUITestTunnelClient.h"
#import <UITestKitCommon/NSURLRequest+SBTUITestTunnelMatch.h>
#import <UITestKitCommon/SBTUITestTunnel.h>
#import <XCTest/XCTest.h>
#include <arpa/inet.h>
#include <ifaddrs.h>

const NSString *SBTUITunnelJsonMimeType = @"application/json";
#define kSBTUITestTunnelErrorDomain @"com.subito.sbtuitesttunnel.error"


@implementation SBTUITestTunnelClient

static NSTimeInterval SBTUITunneledApplicationDefaultTimeout = 30.0;

- (instancetype)initWithApplication:(XCUIApplication *)application
{
    self = [super init];
    
    if (self) {
        _initialLaunchArguments = application.launchArguments;
        _initialLaunchEnvironment = application.launchEnvironment;
        _application = application;
        _userInterfaceAnimationsEnabled = YES;
        _userInterfaceAnimationSpeed = 1;
        
        [self resetInternalState];
    }
    
    return self;
}

- (void)resetInternalState
{
    [self.bonjourBrowser stop];

    self.application.launchArguments = self.initialLaunchArguments;
    self.application.launchEnvironment = self.initialLaunchEnvironment;

    self.startupBlock = nil;

    self.bonjourName = [NSString stringWithFormat:@"com.subito.test.%d.%.0f", [NSProcessInfo processInfo].processIdentifier, (double)(CFAbsoluteTimeGetCurrent() * 100000)];
    self.bonjourBrowser = [[NSNetService alloc] initWithDomain:@"local." type:@"_http._tcp." name:self.bonjourName];
    self.bonjourBrowser.delegate = self;
    self.connected = NO;
    self.connectionPort = 0;
    self.connectionTimeout = SBTUITunneledApplicationDefaultTimeout;
}

- (void)shutDownWithError:(nullable NSError *)error
{
    [self sendSynchronousRequestWithPath:SBTUITunneledApplicationCommandShutDown params:nil assertOnError:NO];
    
    [self resetInternalState];

    if ([self.delegate respondsToSelector:@selector(testTunnelClient:didShutdownWithError:)]) {
        [self.delegate testTunnelClient:self didShutdownWithError:error];
    }
}

- (void)launchTunnel
{
    [self launchTunnelWithStartupBlock:nil];
}

- (void)launchTunnelWithStartupBlock:(void (^)(void))startupBlock
{
    NSMutableArray *launchArguments = [self.application.launchArguments mutableCopy];
    [launchArguments addObject:SBTUITunneledApplicationLaunchSignal];

    if (startupBlock) {
        [launchArguments addObject:SBTUITunneledApplicationLaunchOptionHasStartupCommands];
    }

    self.startupBlock = startupBlock;
    self.application.launchArguments = launchArguments;

    NSMutableDictionary<NSString *, NSString *> *launchEnvironment = [self.application.launchEnvironment mutableCopy];
    launchEnvironment[SBTUITunneledApplicationLaunchEnvironmentBonjourNameKey] = self.bonjourName;

    self.application.launchEnvironment = launchEnvironment;
    
    NSLog(@"[SBTUITestTunnel] Resolving bonjour service %@", self.bonjourName);
    [self.bonjourBrowser resolveWithTimeout:self.connectionTimeout];
    
    [self.delegate testTunnelClientIsReadyToLaunch:self];
    
    [self waitForAppReady];
}

- (void)waitForAppReady
{
    const int timeout = self.connectionTimeout;
    int i = 0;
    for (i = 0; i < timeout; i++) {
        if ([self isAppCruising]) {
            return;
        }
        [NSRunLoop.currentRunLoop runUntilDate:[NSDate dateWithTimeIntervalSinceNow:1.0]];
    }

    NSError *error = [self.class errorWithCode:SBTUITestTunnelErrorLaunchFailed
                                       message:@"Failed waiting for app to be ready"];
    [self shutDownWithError:error];
}

#pragma mark - Bonjour

- (void)netServiceDidResolveAddress:(NSNetService *)service;
{
    if ([service.name isEqualToString:self.bonjourName] && !self.connected) {
        NSAssert(service.port > 0, @"[SBTUITestTunnel] unexpected port 0!");
        
        self.connected = YES;
        
        NSLog(@"[SBTUITestTunnel] Tunnel established on port %ld", (unsigned long)service.port);
        self.connectionPort = service.port;
        
        if (self.startupBlock) {
            self.startupBlock(); // this will eventually add some commands in the startup command queue
        }
        
        [self sendSynchronousRequestWithPath:SBTUITunneledApplicationCommandStartupCommandsCompleted params:@{}];
    }
}

- (void)netService:(NSNetService *)sender didNotResolve:(NSDictionary<NSString *,NSNumber *> *)errorDict
{
    if (!self.connected || ![sender.name isEqualToString:self.bonjourName]) {
        return;
    }

    NSString *message = [NSString localizedStringWithFormat:@"[SBTUITestTunnel] Failed to connect to client app %@", errorDict];
    NSError *error = [self.class errorWithCode:SBTUITestTunnelErrorConnectionToApplicationFailed
                                       message:message];
    [self shutDownWithError:error];
}

#pragma mark - Timeout

+ (void)setConnectionTimeout:(NSTimeInterval)timeout
{
    NSAssert(timeout > 5.0, @"[SBTUITestTunnel] Timeout too short!");
    SBTUITunneledApplicationDefaultTimeout = timeout;
}

#pragma mark - Ping Command

- (BOOL)ping
{
    return [[self sendSynchronousRequestWithPath:SBTUITunneledApplicationCommandPing params:nil assertOnError:NO] isEqualToString:@"YES"];
}

#pragma mark - Quit Command

#pragma mark - Ready Command

- (BOOL)isAppCruising
{
    return [[self sendSynchronousRequestWithPath:SBTUITunneledApplicationCommandCruising params:nil] isEqualToString:@"YES"];
}

#pragma mark - Stub Commands

#pragma mark - Stub And Remove Commands

- (NSString *)stubRequestsMatching:(SBTRequestMatch *)match response:(SBTStubResponse *)response removeAfterIterations:(NSUInteger)iterations
{
    NSDictionary<NSString *, NSString *> *params = @{SBTUITunnelStubMatchRuleKey: [self base64SerializeObject:match],
                                                     SBTUITunnelStubResponseKey: [self base64SerializeObject:response],
                                                     SBTUITunnelStubIterationsKey: [@(iterations) stringValue]
                                                     };

    return [self sendSynchronousRequestWithPath:SBTUITunneledApplicationCommandStubAndRemoveMatching params:params];
}

#pragma mark - Stub Remove Commands

//
- (BOOL)stubRequestsRemoveAll
{
    return [[self sendSynchronousRequestWithPath:SBTUITunneledApplicationCommandStubRequestsRemoveAll params:nil] boolValue];
}

#pragma mark - Rewrite Commands


#pragma mark - Rewrite Remove Commands


#pragma mark - Monitor Requests Commands

- (NSArray<SBTMonitoredNetworkRequest *> *)monitoredRequestsPeekAll
{
    NSString *objectBase64 = [self sendSynchronousRequestWithPath:SBTUITunneledApplicationCommandMonitorPeek params:nil];
    if (objectBase64) {
        NSData *objectData = [[NSData alloc] initWithBase64EncodedString:objectBase64 options:0];

        return [NSKeyedUnarchiver unarchiveObjectWithData:objectData] ?: @[];
    }

    return nil;
}

#pragma mark - Asynchronously Wait for Requests Commands

- (void)waitForMonitoredRequestsMatching:(SBTRequestMatch *)match timeout:(NSTimeInterval)timeout completionBlock:(void (^)(BOOL timeout))completionBlock;
{
    [self waitForMonitoredRequestsMatching:match timeout:timeout iterations:1 completionBlock:completionBlock];
}

- (void)waitForMonitoredRequestsMatching:(SBTRequestMatch *)match timeout:(NSTimeInterval)timeout iterations:(NSUInteger)iterations completionBlock:(void (^)(BOOL timeout))completionBlock;
{
    [self waitForMonitoredRequestsWithMatchingBlock:^BOOL(SBTMonitoredNetworkRequest *request) {
        return [request matches:match];
    } timeout:timeout iterations:iterations completionBlock:completionBlock];
}

- (void)waitForMonitoredRequestsWithMatchingBlock:(BOOL(^)(SBTMonitoredNetworkRequest *))matchingBlock timeout:(NSTimeInterval)timeout iterations:(NSUInteger)iterations completionBlock:(void (^)(BOOL))completionBlock
{
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0), ^() {
        NSTimeInterval start = CFAbsoluteTimeGetCurrent();
        
        BOOL timedout = NO;
        
        for(;;) {
            NSUInteger localIterations = iterations;
            NSArray<SBTMonitoredNetworkRequest *> *requests = [self monitoredRequestsPeekAll];
            
            for (SBTMonitoredNetworkRequest *request in requests) {
                if (matchingBlock(request)) {
                    if (--localIterations == 0) {
                        break;
                    }
                }
            }
            
            if (localIterations < 1) {
                break;
            } else if (CFAbsoluteTimeGetCurrent() - start > timeout) {
                timedout = YES;
                break;
            }
            
            [NSThread sleepForTimeInterval:0.5];
        }
        
        if (completionBlock) {
            completionBlock(timedout);
        }
    });
}

#pragma mark - Synchronously Wait for Requests Commands

#pragma mark - Throttle Requests Commands


#pragma mark - Cookie Block Requests Commands


#pragma mark - NSUserDefaults Commands

#pragma mark - NSBundle

#pragma mark - Copy Commands

- (BOOL)uploadItemAtPath:(NSString *)srcPath toPath:(NSString *)destPath relativeTo:(NSSearchPathDirectory)baseFolder
{
    NSAssert(![srcPath hasPrefix:@"file:"], @"Call this methon passing srcPath using [NSURL path] not [NSURL absoluteString]!");

    NSData *data = [NSData dataWithContentsOfURL:[NSURL fileURLWithPath:srcPath]];

    if (!data) {
        return NO;
    }

    NSDictionary<NSString *, NSString *> *params = @{SBTUITunnelUploadDataKey: [self base64SerializeData:data],
                                                     SBTUITunnelUploadDestPathKey: [self base64SerializeObject:destPath ?: @""],
                                                     SBTUITunnelUploadBasePathKey: [@(baseFolder) stringValue]};

    return [[self sendSynchronousRequestWithPath:SBTUITunneledApplicationCommandUploadData params:params] boolValue];
}

- (NSArray<NSData *> *)downloadItemsFromPath:(NSString *)path relativeTo:(NSSearchPathDirectory)baseFolder
{
    NSDictionary<NSString *, NSString *> *params = @{SBTUITunnelDownloadPathKey: [self base64SerializeObject:path ?: @""],
                                                     SBTUITunnelDownloadBasePathKey: [@(baseFolder) stringValue]};

    NSString *itemsBase64 = [self sendSynchronousRequestWithPath:SBTUITunneledApplicationCommandDownloadData params:params];

    if (itemsBase64) {
        NSData *itemsData = [[NSData alloc] initWithBase64EncodedString:itemsBase64 options:0];

        return [NSKeyedUnarchiver unarchiveObjectWithData:itemsData];
    }

    return nil;
}

#pragma mark - Custom Commands

#pragma mark - Other Commands

- (BOOL)setUserInterfaceAnimationsEnabled:(BOOL)enabled
{
    _userInterfaceAnimationsEnabled = enabled;
    
    NSDictionary<NSString *, NSString *> *params = @{SBTUITunnelObjectKey: [@(enabled) stringValue]};
    
    return [[self sendSynchronousRequestWithPath:SBTUITunneledApplicationCommandSetUserInterfaceAnimations params:params] boolValue];
}

- (BOOL)userInterfaceAnimationsEnabled
{
    return _userInterfaceAnimationsEnabled;
}

- (BOOL)setUserInterfaceAnimationSpeed:(NSInteger)speed
{
    _userInterfaceAnimationSpeed = speed;
    
    NSDictionary<NSString *, NSString *> *params = @{SBTUITunnelObjectKey: [@(speed) stringValue]};
    
    return [[self sendSynchronousRequestWithPath:SBTUITunneledApplicationCommandSetUserInterfaceAnimationSpeed params:params] boolValue];
}

- (NSInteger)userInterfaceAnimationSpeed
{
    return _userInterfaceAnimationSpeed;
}

#pragma mark - XCUITest scroll extensions

- (BOOL)scrollTableViewWithIdentifier:(nonnull NSString *)identifier toRow:(NSInteger)row animated:(BOOL)flag
{
    NSAssert([identifier length] > 0, @"Invalid empty identifier!");
    
    NSDictionary<NSString *, NSString *> *params = @{SBTUITunnelObjectKey: identifier,
                                                     SBTUITunnelObjectValueKey: [@(row) stringValue],
                                                     SBTUITunnelObjectAnimatedKey: [@(flag) stringValue]};
    
    return [[self sendSynchronousRequestWithPath:SBTUITunneledApplicationCommandXCUIExtensionScrollTableView params:params] boolValue];
}

- (BOOL)scrollCollectionViewWithIdentifier:(nonnull NSString *)identifier toRow:(NSInteger)row animated:(BOOL)flag
{
    NSAssert([identifier length] > 0, @"Invalid empty identifier!");
    
    NSDictionary<NSString *, NSString *> *params = @{SBTUITunnelObjectKey: identifier,
                                                     SBTUITunnelObjectValueKey: [@(row) stringValue],
                                                     SBTUITunnelObjectAnimatedKey: [@(flag) stringValue]};
    
    return [[self sendSynchronousRequestWithPath:SBTUITunneledApplicationCommandXCUIExtensionScrollCollectionView params:params] boolValue];
}

- (BOOL)scrollScrollViewWithIdentifier:(nonnull NSString *)identifier toElementWitIdentifier:(nonnull NSString *)targetIdentifier animated:(BOOL)flag
{
    NSAssert([identifier length] > 0, @"Invalid empty identifier!");
    NSAssert([targetIdentifier length] > 0, @"Invalid empty target identifier!");
    
    NSDictionary<NSString *, NSString *> *params = @{SBTUITunnelObjectKey: identifier,
                                                     SBTUITunnelObjectValueKey: targetIdentifier,
                                                     SBTUITunnelObjectAnimatedKey: [@(flag) stringValue]};
    
    return [[self sendSynchronousRequestWithPath:SBTUITunneledApplicationCommandXCUIExtensionScrollScrollView params:params] boolValue];
}

#pragma mark - XCUITest 3D touch extensions

- (BOOL)forcePressViewWithIdentifier:(nonnull NSString *)identifier
{
    NSAssert([identifier length] > 0, @"Invalid empty identifier!");
    
    NSDictionary<NSString *, NSString *> *params = @{SBTUITunnelObjectKey: identifier};
    
    return [[self sendSynchronousRequestWithPath:SBTUITunneledApplicationCommandXCUIExtensionForceTouchView params:params] boolValue];
}

#pragma mark - XCUITest CLLocation extensions

- (BOOL)coreLocationStubEnabled:(BOOL)flag
{
    NSDictionary<NSString *, NSString *> *params = @{SBTUITunnelObjectValueKey: flag ? @"YES" : @"NO"};
    
    return [[self sendSynchronousRequestWithPath:SBTUITunneledApplicationCommandCoreLocationStubbing params:params] boolValue];
}

- (BOOL)coreLocationStubAuthorizationStatus:(CLAuthorizationStatus)status
{
    NSDictionary<NSString *, NSString *> *params = @{SBTUITunnelObjectValueKey: [@(status) stringValue]};
    
    return [[self sendSynchronousRequestWithPath:SBTUITunneledApplicationCommandCoreLocationStubAuthorizationStatus params:params] boolValue];
}

- (BOOL)coreLocationStubLocationServicesEnabled:(BOOL)flag
{
    NSDictionary<NSString *, NSString *> *params = @{SBTUITunnelObjectValueKey: flag ? @"YES" : @"NO"};
    
    return [[self sendSynchronousRequestWithPath:SBTUITunneledApplicationCommandCoreLocationStubServiceStatus params:params] boolValue];
}

- (BOOL)coreLocationNotifyLocationUpdate:(nonnull NSArray<CLLocation *>*)locations
{
    NSAssert([locations count] > 0, @"Location array should contain at least one element!");
    
    NSDictionary<NSString *, NSString *> *params = @{SBTUITunnelObjectKey: [self base64SerializeObject:locations]};
    
    return [[self sendSynchronousRequestWithPath:SBTUITunneledApplicationCommandCoreLocationNotifyUpdate params:params] boolValue];
}

- (BOOL)coreLocationNotifyLocationError:(nonnull NSError *)error
{
    NSDictionary<NSString *, NSString *> *params = @{SBTUITunnelObjectKey: [self base64SerializeObject:error]};
    
    return [[self sendSynchronousRequestWithPath:SBTUITunneledApplicationCommandCoreLocationNotifyFailure params:params] boolValue];
}

#pragma mark - XCUITest UNUserNotificationCenter extensions

- (BOOL)notificationCenterStubEnabled:(BOOL)flag API_AVAILABLE(ios(10))
{
    NSDictionary<NSString *, NSString *> *params = @{SBTUITunnelObjectValueKey: flag ? @"YES" : @"NO"};
    
    return [[self sendSynchronousRequestWithPath:SBTUITunneledApplicationCommandNotificationCenterStubbing params:params] boolValue];
}

- (BOOL)notificationCenterStubAuthorizationStatus:(UNAuthorizationStatus)status API_AVAILABLE(ios(10))
{
    NSDictionary<NSString *, NSString *> *params = @{SBTUITunnelObjectValueKey: [@(status) stringValue]};
    
    return [[self sendSynchronousRequestWithPath:SBTUITunneledApplicationCommandNotificationCenterStubAuthorizationStatus params:params] boolValue];
}


#pragma mark - Helper Methods

- (NSString *)base64SerializeObject:(id)obj
{
    NSData *objData = [NSKeyedArchiver archivedDataWithRootObject:obj];
    
    return [self base64SerializeData:objData];
}

- (NSString *)base64SerializeData:(NSData *)data
{
    if (!data) {
        NSError *error = [self.class errorWithCode:SBTUITestTunnelErrorOtherFailure
                                           message:@"[SBTUITestTunnel] Failed to serialize object"];
        [self shutDownWithError:error];
        return @"";
    } else {
        return [[data base64EncodedStringWithOptions:0] stringByAddingPercentEncodingWithAllowedCharacters:[NSCharacterSet alphanumericCharacterSet]];
    }
}

- (NSString *)sendSynchronousRequestWithPath:(NSString *)path params:(NSDictionary<NSString *, NSString *> *)params assertOnError:(BOOL)assertOnError
{
    if (self.connectionlessBlock) {
        if ([NSThread isMainThread]) {
            return self.connectionlessBlock(path, params);
        } else {
            __block NSString *ret = @"";
            __weak typeof(self)weakSelf = self;
            dispatch_sync(dispatch_get_main_queue(), ^{
                ret = weakSelf.connectionlessBlock(path, params);
            });
            return ret;
        }
    }
    
    if (self.connectionPort == 0) {
        return nil; // connection still not established
    }
    
    NSString *urlString = [NSString stringWithFormat:@"http://%@:%d/%@", SBTUITunneledApplicationDefaultHost, (unsigned int)self.connectionPort, path];
    
    NSURL *url = [NSURL URLWithString:urlString];
    
    NSMutableURLRequest *request = nil;
    NSURLComponents *components = [NSURLComponents componentsWithURL:url resolvingAgainstBaseURL:NO];
    
    NSMutableArray *queryItems = [NSMutableArray array];
    [params enumerateKeysAndObjectsUsingBlock:^(NSString *key, NSString *value, BOOL *stop) {
        [queryItems addObject:[NSURLQueryItem queryItemWithName:key value:value]];
    }];
    components.queryItems = queryItems;
    
    if ([SBTUITunnelHTTPMethod isEqualToString:@"GET"]) {
        request = [NSMutableURLRequest requestWithURL:components.URL];
    } else if  ([SBTUITunnelHTTPMethod isEqualToString:@"POST"]) {
        request = [NSMutableURLRequest requestWithURL:url];
        
        request.HTTPBody = [components.query dataUsingEncoding:NSUTF8StringEncoding];
    }
    request.HTTPMethod = SBTUITunnelHTTPMethod;
    
    if (!request) {
        NSError *error = [self.class errorWithCode:SBTUITestTunnelErrorOtherFailure
                                           message:@"[SBTUITestTunnel] Did fail to create url component"];
        [self shutDownWithError:error];
        return nil;
    }
    
    dispatch_semaphore_t synchRequestSemaphore = dispatch_semaphore_create(0);
    
    NSURLSession *session = [NSURLSession sharedSession];
    __block NSString *responseId = nil;
    
    [[session dataTaskWithRequest:request completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
        if (error.code == -1022) {
            NSAssert(NO, @"Check that ATS security policy is properly setup, refer to documentation");
        }
        
        if (![response isKindOfClass:[NSHTTPURLResponse class]]) {
            if (assertOnError) {
                NSLog(@"[SBTUITestTunnel] Failed to get http response: %@", request);
                // [weakSelf terminate];
            }
        } else {
            NSDictionary *jsonData = [NSJSONSerialization JSONObjectWithData:data options:NSJSONReadingMutableContainers error:nil];
            responseId = jsonData[SBTUITunnelResponseResultKey];
            
            if (assertOnError) {
                if (((NSHTTPURLResponse *)response).statusCode != 200) {
                    NSLog(@"[SBTUITestTunnel] Message sending failed: %@", request);
                }
            }
        }
        
        dispatch_semaphore_signal(synchRequestSemaphore);
    }] resume];
    
    dispatch_semaphore_wait(synchRequestSemaphore, DISPATCH_TIME_FOREVER);
    
    return responseId;
}

- (NSString *)sendSynchronousRequestWithPath:(NSString *)path params:(NSDictionary<NSString *, NSString *> *)params
{
    return [self sendSynchronousRequestWithPath:path params:params assertOnError:YES];
}

#pragma mark - Error Helpers

+ (NSError *)errorWithCode:(SBTUITestTunnelError)code message:(NSString *)message
{
    return [NSError errorWithDomain:kSBTUITestTunnelErrorDomain
                               code:code
                           userInfo:@{ NSLocalizedDescriptionKey : message }];
}

@end

#endif

File: Pod/Client/SBTUITestTunnelClient.swift
//
//  Copyright © 2019 Qonto. All rights reserved.
//

import Foundation
import UITestKitCommon
import XCTest

let SBTUITunnelJsonMimeType = "application/json"

let kSBTUITestTunnelErrorDomain = "com.subito.sbtuitesttunnel.error"

@objc public class SBTUITestTunnelClient: NSObject, NetServiceDelegate {
    var userInterfaceAnimationsEnabled = false
    var userInterfaceAnimationSpeed = 0

    private weak var application: XCUIApplication?
    private var connectionPort = 0
    private var connected = false
    private var connectionTimeout: TimeInterval = 0.0
    private var stubOnceIds: [AnyHashable]?
    private var bonjourName: String = ""
    private var bonjourBrowser: NetService?
    private var startupBlock: (() -> Void)?
    private var initialLaunchArguments: [String] = []
    private var initialLaunchEnvironment: [String: String] = [:]
    var connectionlessBlock: ((String, [String: String]) -> String)?
    private var sbtuiTunneledApplicationDefaultTimeout: TimeInterval = 0.0

    public weak var delegate: SBTUITestTunnelClientDelegate?

    private var SBTUITunneledApplicationDefaultTimeout = 30.0

    @objc public init(application: XCUIApplication) {
        super.init()

        initialLaunchArguments = application.launchArguments
        initialLaunchEnvironment = application.launchEnvironment
        self.application = application
        userInterfaceAnimationsEnabled = true
        userInterfaceAnimationSpeed = 1

        resetInternalState()
    }

    @objc func resetInternalState() {
        bonjourBrowser?.stop()

        application?.launchArguments = initialLaunchArguments
        application?.launchEnvironment = initialLaunchEnvironment

        startupBlock = nil

        bonjourName = String(format: "com.subito.test.%d.%.0f", ProcessInfo.processInfo.processIdentifier, Double(CFAbsoluteTimeGetCurrent() * 100_000))
        bonjourBrowser = NetService(domain: "local.", type: "_http._tcp.", name: bonjourName)
        bonjourBrowser?.delegate = self
        connected = false
        connectionPort = 0
        connectionTimeout = SBTUITunneledApplicationDefaultTimeout
    }

    @objc func shutDownWithError(_ error: Error?) {
        sendSynchronousRequest(withPath: SBTUITunneledApplicationCommandShutDown, params: nil, assertOnError: false)

        resetInternalState()

        if delegate?.responds(to: #selector(delegate?.testTunnelClient(_:didShutdownWithError:))) ?? true {
            delegate?.testTunnelClient?(self, didShutdownWithError: error)
        }
    }

    @objc public func launchTunnel() {
        launchTunnel(withStartupBlock: nil)
    }

    @objc public func launchTunnel(withStartupBlock startupBlock: (() -> Void)? = nil) {
        var launchArguments: [String] = application?.launchArguments ?? []
        launchArguments.append(SBTUITunneledApplicationLaunchSignal)

        if startupBlock != nil {
            launchArguments.append(SBTUITunneledApplicationLaunchOptionHasStartupCommands)
        }

        self.startupBlock = startupBlock
        application?.launchArguments = launchArguments

        var launchEnvironment = application?.launchEnvironment
        launchEnvironment?[SBTUITunneledApplicationLaunchEnvironmentBonjourNameKey] = bonjourName

        application?.launchEnvironment = launchEnvironment ?? [:]

        print("[SBTUITestTunnel] Resolving bonjour service \(bonjourName)")
        bonjourBrowser?.resolve(withTimeout: connectionTimeout)

        delegate?.testTunnelClientIsReady(toLaunch: self)

        waitForAppReady()
    }

    @objc func setConnectionTimeout(_ timeout: TimeInterval) {
        assert(timeout > 5.0, "[SBTUITestTunnel] Timeout too short!")
        SBTUITunneledApplicationDefaultTimeout = timeout
    }

    @objc func waitForAppReady() {
        let timeout = Int(connectionTimeout)

        for _ in 0 ..< timeout {
            if isAppCruising() {
                return
            }
            RunLoop.current.run(until: Date(timeIntervalSinceNow: 1.0))
        }

        let error = self.error(withCode: .launchFailed, message: "Failed waiting for app to be ready")
        shutDownWithError(error)
    }

    @objc public func netServiceDidResolveAddress(_ service: NetService) {
        if service.name == bonjourName, !connected {
            assert(service.port > 0, "[SBTUITestTunnel] unexpected port 0!")

            connected = true

            print(String(format: "[SBTUITestTunnel] Tunnel established on port %ld", UInt(service.port)))
            connectionPort = service.port

            if startupBlock != nil {
                startupBlock!() // this will eventually add some commands in the startup command queue
            }

            sendSynchronousRequest(withPath: SBTUITunneledApplicationCommandStartupCommandsCompleted, params: [:])
        }
    }

    @objc public func netService(_ sender: NetService, didNotResolve errorDict: [String: NSNumber]) {
        if !connected || !(sender.name == bonjourName) {
            return
        }

        let message = "[SBTUITestTunnel] Failed to connect to client app \(errorDict)"
        let error = self.error(withCode: .connectionToApplicationFailed, message: message)
        shutDownWithError(error)
    }

    @discardableResult
    @objc func sendSynchronousRequest(withPath path: String, params: [String: String]? = [:], assertOnError: Bool) -> String? {
        let options = params ?? [:]

        if connectionlessBlock != nil {
            if Thread.isMainThread {
                return connectionlessBlock!(path, options)
            } else {
                var ret = ""
                weak var weakSelf = self
                DispatchQueue.main.sync {
                    guard let item = weakSelf?.connectionlessBlock else {
                        return
                    }
                    ret = item(path, options)
                }
                return ret
            }
        }

        guard connectionPort != 0 else {
            return nil // connection still not established
        }

        let urlString = "http://\(SBTUITunneledApplicationDefaultHost):\(UInt(connectionPort))/\(path)"

        let url = URL(string: urlString)

        var request: URLRequest?
        var components: URLComponents?
        if let url = url {
            components = URLComponents(url: url, resolvingAgainstBaseURL: false)
        }

        var queryItems: [AnyHashable] = []
        (params as NSDictionary?)?.enumerateKeysAndObjects { key, value, _ in
            let keyItem = key as? String ?? ""
            let valueItem = value as? String ?? ""
            queryItems.append(URLQueryItem(name: keyItem, value: valueItem))
        }
        components?.queryItems = queryItems as? [URLQueryItem]

        if SBTUITunnelHTTPMethod.isEqual("GET") {
            if let URL = components?.url {
                request = URLRequest(url: URL)
            }
        } else if SBTUITunnelHTTPMethod.isEqual("POST") {
            if let url = url {
                request = URLRequest(url: url)
            }

            request?.httpBody = components?.query?.data(using: .utf8)
        }
        request?.httpMethod = SBTUITunnelHTTPMethod

        guard let urlRequest = request else {
            let error = self.error(withCode: .otherFailure, message: "[SBTUITestTunnel] Did fail to create url component")
            shutDownWithError(error)
            return nil
        }

        let synchRequestSemaphore = DispatchSemaphore(value: 0)

        let session = URLSession.shared
        var responseId: String?

        (session.dataTask(with: urlRequest, completionHandler: { data, response, error in
            if (error as NSError?)?.code == -1022 {
                assert(false, "Check that ATS security policy is properly setup, refer to documentation")
            }

            if !(response is HTTPURLResponse) {
                if assertOnError {
                    print("[SBTUITestTunnel] Failed to get http response: \(String(describing: request))")
                    // [weakSelf terminate];
                }
            } else {
                var jsonData: [AnyHashable: Any]?
                do {
                    if let data = data {
                        jsonData = try JSONSerialization.jsonObject(with: data, options: .mutableContainers) as? [AnyHashable: Any]
                    }
                } catch {}
                responseId = jsonData?[SBTUITunnelResponseResultKey] as? String

                if assertOnError {
                    if (response as? HTTPURLResponse)?.statusCode != 200 {
                        print("[SBTUITestTunnel] Message sending failed: \(String(describing: request))")
                    }
                }
            }

            synchRequestSemaphore.signal()
        })).resume()

        _ = synchRequestSemaphore.wait(timeout: DispatchTime.distantFuture) == .success

        return responseId
    }

    @discardableResult
    @objc func sendSynchronousRequest(withPath path: String, params: [String: String]?) -> String? {
        return sendSynchronousRequest(withPath: path, params: params, assertOnError: true)
    }
}
